Tests/GRPCHTTP2CoreTests/Client/Connection/LoadBalancers/SubchannelTests.swift

```diff
@@ -0,0 +1,340 @@
+          try await stream.execute { inbound, outbound in
+            try await outbound.write(.metadata([:]))
+            try await outbound.write(.message([0, 1, 2]))
+            outbound.finish()
+
+            for try await part in inbound {
+              switch part {
+              case .metadata:
+                ()  // Don't validate, contains http/2 specific metadata too.
+              case .message(let message):
+                XCTAssertEqual(message, [0, 1, 2])
+              case .status(let status, _):
+                XCTAssertEqual(status.code, .ok)
+                XCTAssertEqual(status.message, "")
+              }
+            }
+          }
+          subchannel.close()
+
+        default:
+          ()
+        }
+      }
+
+      group.cancelAll()
+    }
+  }
+
+  func testConnectEventuallySucceeds() async throws {
+    let path = "test-connect-eventually-succeeds"
+    let subchannel = self.makeSubchannel(
+      address: .unixDomainSocket(path: path),
+      connector: .posix(),
+      backoff: .fixed(at: .milliseconds(100))
+    )
+
+    await withThrowingTaskGroup(of: Void.self) { group in
+      group.addTask { await subchannel.run() }
+
+      var hasServer = false
+      var events = [Subchannel.Event]()
+
+      for await event in subchannel.events {
+        events.append(event)
+        switch event {
+        case .connectivityStateChanged(.idle):
+          subchannel.connect()
+
+        case .connectivityStateChanged(.transientFailure):
+          // Don't start more than one server.
+          if hasServer { continue }
+          hasServer = true
+
+          group.addTask {
+            let server = TestServer(eventLoopGroup: .singletonMultiThreadedEventLoopGroup)
+            _ = try await server.bind(to: .uds(path))
+            try await server.run { _, _ in
+              XCTFail("Unexpected stream")
+            }
+          }
+
+        case .connectivityStateChanged(.ready):
+          subchannel.close()
+
+        case .connectivityStateChanged(.shutdown):
+          group.cancelAll()
+
+        default:
+          ()
+        }
+      }
+
+      // First four events are known:
+      XCTAssertEqual(
+        Array(events.prefix(4)),
+        [
+          .connectivityStateChanged(.idle),
+          .connectivityStateChanged(.connecting),
+          .connectivityStateChanged(.transientFailure),
+          .connectivityStateChanged(.connecting),
+        ]
+      )
+
+      // Because there is backoff timing involved, the subchannel may flip from transient failure
+      // to connecting multiple times. Just check that it eventually becomes ready and is then
+      // shutdown.
+      XCTAssertEqual(
+        Array(events.suffix(2)),
+        [
+          .connectivityStateChanged(.ready),
+          .connectivityStateChanged(.shutdown),
+        ]
+      )
+    }
+  }
+
+  func testConnectIteratesThroughAddresses() async throws {
```

Comment:
Could we add an additional test to make sure we wrap around the list of addresses?

Reply:
Yeah, good call, added `testConnectIteratesThroughAddressesWithBackoff`